File: Election.R
# Do install.packages("readr") if you don't have the package installed
library(dplyr)
library(readr)
library(tidyr)

# Read the election data from the CSV file
election <- read_csv("ElectionResults.csv")

#Change rank to integer
election <- election %>%
  mutate(R = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", R))) %>%
  mutate(D = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", D))) %>%
  mutate(L = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", L))) %>%
  mutate(G = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", G))) %>%
  mutate(Adithi = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", Adithi))) %>%
  mutate(Srinjoy = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", Srinjoy))) %>%
  mutate(Tracy = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", Tracy))) %>%
  mutate(Hunter = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", Hunter))) %>%
  mutate(Archith = as.integer(gsub("([0-9]+)(st|nd|rd|th)", "\\1", Archith)))
# Define isSenior and MF columns based on wing information
election <- election %>%
  mutate(
    isSenior = case_when(
      Wing %in% c("C100", "A100", "D100A", "D100B", "D200A", "D200B", "D300A", "D300B") ~ TRUE,
      Wing %in% c("A200", "C200A", "C200B", "C300A", "C300B", "B200", "A300", "B300") ~ FALSE,
      Wing == "Other" ~ sample(c(TRUE, FALSE), size = n(), replace = TRUE)
    ),
    MF = case_when(
      Wing %in% c("A100", "C200A", "C200B", "B200", "A200", "D200A", "D200B") ~ "F",
      Wing %in% c("C100", "D100A", "D100B", "A300", "B300", "C300A", "C300B", "D300A", "D300B") ~ "M",
      Wing == "Other" ~ sample(c("M", "F"), size = n(), replace = TRUE)
    )
  )

# Summarize counts of seniors, juniors, guys, and girls
summary_counts <- election %>%
  summarise(
    num_seniors = sum(isSenior == TRUE, na.rm = TRUE),
    num_juniors = sum(isSenior == FALSE, na.rm = TRUE),
    num_guys = sum(MF == "M", na.rm = TRUE),
    num_girls = sum(MF == "F", na.rm = TRUE)
  )
num_seniors = summary_counts$num_seniors
num_juniors = summary_counts$num_juniors
num_guys = summary_counts$num_guys
num_girls = summary_counts$num_girls

issues_long <- election %>%
  separate_rows(Issues, sep = ",\\s*") %>%  # Split issues by comma and remove extra spaces
  mutate(Issues = trimws(Issues))  # Trim any leading/trailing whitespace from issue names

issue_counts_class <- issues_long %>%
  group_by(Issues) %>%
  summarise(
    Senior_Frequency = round(100 * sum(isSenior == TRUE, na.rm = TRUE) / num_seniors, 1),   # Count of seniors
    Junior_Frequency = round(100 * sum(isSenior == FALSE, na.rm = TRUE) / num_juniors, 1),  # Count of juniors
    .groups = "drop" 
  )
# Count the occurrences of each issue by gender (MF)
issue_counts_gender <- issues_long %>%
  group_by(Issues) %>%
  summarise(
    Guys_Frequency = round(100 * sum(MF == "M", na.rm = TRUE) / num_guys, 1),   # Count of males
    Girls_Frequency = round(100 * sum(MF == "F", na.rm = TRUE) / num_girls, 1), # Count of females
    .groups = "drop" 
  )


# Print the issue counts
View(issue_counts_class)
View(issue_counts_gender)

# Reshape the party ranking columns (R, D, L, G) from wide to long format
ranking_long <- election %>%
  pivot_longer(cols = c(R, D, L, G), 
               names_to = "party", 
               values_to = "rank") %>%
  filter(!is.na(rank))  # Remove rows where rank is NA

# Step 1: Analyze average rank for each party by Senior vs. Junior and Guys vs. Girls

# Senior vs. Junior average ranks
senior_junior_rankings <- ranking_long %>%
  group_by(isSenior, party) %>%
  summarise(avg_rank = mean(rank, na.rm = TRUE), .groups = "drop") %>%
  arrange(isSenior, avg_rank)

# Guys vs. Girls average ranks
gender_rankings <- ranking_long %>%
  group_by(MF, party) %>%
  summarise(avg_rank = mean(rank, na.rm = TRUE), .groups = "drop") %>%
  arrange(MF, avg_rank)

# Step 2: Analyze FPTP election choices for President and Senator

# NOTE: To replicate for the other elections, simply change the cols variable in the code below to
# include the other election positions (e.g., "President", "Senator", "Texas_House64", etc.)
# Then, in the reformat step, change the filter for the position (names_to) to the desired election position.
# Finally, rename the tables accordingly for clarity in the output.

# Senior vs. Junior FPTP choices
senior_junior_fptp <- election %>%
  pivot_longer(cols = c(President, Senator), 
               names_to = "position", 
               values_to = "choice") %>%
  group_by(position, isSenior, choice) %>%
  summarise(vote_count = n(), .groups = "drop") %>%
  arrange(position, isSenior, desc(vote_count))

# Guys vs. Girls FPTP choices
gender_fptp <- election %>%
  pivot_longer(cols = c(President, Senator), 
               names_to = "position", 
               values_to = "choice") %>%
  group_by(position, MF, choice) %>%
  summarise(vote_count = n(), .groups = "drop") %>%
  arrange(position, MF, desc(vote_count))

# Reformat into tables for display (Senior vs. Junior President election)
presidential_fptp_class <- senior_junior_fptp %>%
  filter(position == "President") %>%  # Filter for Presidential position only
  group_by(choice, isSenior) %>%        # Group by candidate and Senior status
  summarise(vote_count = sum(vote_count, na.rm = TRUE), .groups = "drop") %>%
  spread(key = isSenior, value = vote_count, fill = 0) %>%  # Spread to wide format
  rename(
    Senior = `TRUE`,   # Rename for clarity
    Junior = `FALSE`
  )

# Reformat into tables for display (Guys vs. Girls President election)
presidential_fptp_gender <- gender_fptp %>%
  filter(position == "President") %>%  # Filter for Presidential position only
  group_by(choice, MF) %>%        # Group by candidate and Senior status
  summarise(vote_count = sum(vote_count, na.rm = TRUE), .groups = "drop") %>%
  spread(key = MF, value = vote_count, fill = 0) %>%  # Spread to wide format
  rename(
    Guys = `M`,   # Rename for clarity
    Girls = `F`
  )

# Reformat into tables for display (Senior vs. Junior Senate election)
senate_fptp_class <- senior_junior_fptp %>%
  filter(position == "Senator") %>%  # Filter for Presidential position only
  group_by(choice, isSenior) %>%        # Group by candidate and Senior status
  summarise(vote_count = sum(vote_count, na.rm = TRUE), .groups = "drop") %>%
  spread(key = isSenior, value = vote_count, fill = 0) %>%  # Spread to wide format
  rename(
    Senior = `TRUE`,   # Rename for clarity
    Junior = `FALSE`
  )

# Reformat into tables for display (Guys vs. Girls Senate election)
senate_fptp_gender <- gender_fptp %>%
  filter(position == "Senator") %>%  # Filter for Presidential position only
  group_by(choice, MF) %>%        # Group by candidate and Senior status
  summarise(vote_count = sum(vote_count, na.rm = TRUE), .groups = "drop") %>%
  spread(key = MF, value = vote_count, fill = 0) %>%  # Spread to wide format
  rename(
    Guys = `M`,   # Rename for clarity
    Girls = `F`
  )

# Display results
list(
  "Senior vs Junior Average Party Rankings" = senior_junior_rankings,
  "Guys vs Girls Average Party Rankings" = gender_rankings,
  "Senior vs Junior President Choices" = presidential_fptp_class,
  "Senior vs Junior Senate Choices" = senate_fptp_class,
  "Guys vs Girls President Choices" = presidential_fptp_gender,
  "Senior vs Junior Senate Choices" = senate_fptp_gender
  
)

File: Election.py
import pandas as pd
import numpy as np

# Load the election data
election = pd.read_csv("ElectionResults.csv")

# Convert rank columns to integer
for col in ['R', 'D', 'L', 'G', 'Adithi', 'Srinjoy', 'Tracy', 'Hunter', 'Archith']:
    election[col] = election[col].str.extract(r'(\d+)').astype(float).astype('Int64')

# Define isSenior and MF columns based on wing information
senior_wings = ["C100", "A100", "D100A", "D100B", "D200A", "D200B", "D300A", "D300B"]
junior_wings = ["A200", "C200A", "C200B", "C300A", "C300B", "B200", "A300", "B300"]
male_wings = ["C100", "D100A", "D100B", "A300", "B300", "C300A", "C300B", "D300A", "D300B"]
female_wings = ["A100", "C200A", "C200B", "B200", "A200", "D200A", "D200B"]

# Assign values for isSenior and MF
election['isSenior'] = np.where(election['Wing'].isin(senior_wings), True,
                        np.where(election['Wing'].isin(junior_wings), False,
                                 np.random.choice([True, False], len(election))))

election['MF'] = np.where(election['Wing'].isin(female_wings), 'F',
                 np.where(election['Wing'].isin(male_wings), 'M',
                          np.random.choice(['M', 'F'], len(election))))

# Summarize counts of seniors, juniors, guys, and girls
summary_counts = {
    "num_seniors": election['isSenior'].sum(),
    "num_juniors": len(election) - election['isSenior'].sum(),
    "num_guys": (election['MF'] == 'M').sum(),
    "num_girls": (election['MF'] == 'F').sum()
}
num_seniors = summary_counts['num_seniors']
num_juniors = summary_counts['num_juniors']
num_guys = summary_counts['num_guys']
num_girls = summary_counts['num_girls']

detailed_counts = election.groupby(['isSenior', 'MF']).size().unstack(fill_value=0)

# Rename columns for clarity
detailed_counts = detailed_counts.rename(columns={True: 'Senior', False: 'Junior'}, index={'M': 'Guys', 'F': 'Girls'})

# Flatten the result to create named columns for each group
# Ensure expected labels exist, defaulting to 0 if missing
counts_dict = {
    "Senior_Guys": detailed_counts.loc['Guys', 'Senior'] if ('Guys' in detailed_counts.index and 'Senior' in detailed_counts.columns) else 0,
    "Senior_Girls": detailed_counts.loc['Girls', 'Senior'] if ('Girls' in detailed_counts.index and 'Senior' in detailed_counts.columns) else 0,
    "Junior_Guys": detailed_counts.loc['Guys', 'Junior'] if ('Guys' in detailed_counts.index and 'Junior' in detailed_counts.columns) else 0,
    "Junior_Girls": detailed_counts.loc['Girls', 'Junior'] if ('Girls' in detailed_counts.index and 'Junior' in detailed_counts.columns) else 0
}

# Convert to a DataFrame for easier viewing (optional)
counts_df = pd.DataFrame([counts_dict])

print(counts_df)
# Separate issues into long format
issues_long = election.assign(Issues=election['Issues'].str.split(',')).explode('Issues')
issues_long['Issues'] = issues_long['Issues'].str.strip()

# Count issues by Senior vs Junior
issue_counts_class = issues_long.groupby('Issues').apply(
    lambda x: pd.Series({
        "Senior_Frequency": round(100 * x['isSenior'].sum() / num_seniors, 1),
        "Junior_Frequency": round(100 * (~x['isSenior']).sum() / num_juniors, 1)
    })
).reset_index()

# Count issues by Gender (MF)
issue_counts_gender = issues_long.groupby('Issues').apply(
    lambda x: pd.Series({
        "Guys_Frequency": round(100 * (x['MF'] == 'M').sum() / num_guys, 1),
        "Girls_Frequency": round(100 * (x['MF'] == 'F').sum() / num_girls, 1)
    })
).reset_index()

# Reshape party rankings to long format and calculate average rank for Senior vs Junior and Guys vs Girls
ranking_long = election.melt(
    id_vars=['Wing', 'isSenior', 'MF'], 
    value_vars=['R', 'D', 'L', 'G'], 
    var_name='party', 
    value_name='rank'
).dropna()

# Average rank for Senior vs. Junior
senior_junior_rankings = ranking_long.groupby(['isSenior', 'party']).agg(avg_rank=('rank', 'mean')).reset_index()
senior_junior_rankings = senior_junior_rankings.sort_values(['isSenior', 'avg_rank'])

# Average rank for Guys vs. Girls
gender_rankings = ranking_long.groupby(['MF', 'party']).agg(avg_rank=('rank', 'mean')).reset_index()
gender_rankings = gender_rankings.sort_values(['MF', 'avg_rank'])

# FPTP analysis for President and Senator
fptp_long = election.melt(
    id_vars=['isSenior', 'MF'], 
    value_vars=['President', 'Senator'], 
    var_name='position', 
    value_name='choice'
)

# Senior vs Junior FPTP choices
senior_junior_fptp = fptp_long.groupby(['position', 'isSenior', 'choice']).size().reset_index(name='vote_count')
senior_junior_fptp = senior_junior_fptp.sort_values(['position', 'isSenior', 'vote_count'], ascending=[True, True, False])

# Guys vs Girls FPTP choices
gender_fptp = fptp_long.groupby(['position', 'MF', 'choice']).size().reset_index(name='vote_count')
gender_fptp = gender_fptp.sort_values(['position', 'MF', 'vote_count'], ascending=[True, True, False])

# Reformat FPTP choices to wide format
presidential_fptp_class = senior_junior_fptp[senior_junior_fptp['position'] == 'President'].pivot_table(
    index='choice', columns='isSenior', values='vote_count', fill_value=0
).rename(columns={True: 'Senior', False: 'Junior'}).reset_index()

presidential_fptp_gender = gender_fptp[gender_fptp['position'] == 'President'].pivot_table(
    index='choice', columns='MF', values='vote_count', fill_value=0
).rename(columns={'M': 'Guys', 'F': 'Girls'}).reset_index()

senate_fptp_class = senior_junior_fptp[senior_junior_fptp['position'] == 'Senator'].pivot_table(
    index='choice', columns='isSenior', values='vote_count', fill_value=0
).rename(columns={True: 'Senior', False: 'Junior'}).reset_index()

senate_fptp_gender = gender_fptp[gender_fptp['position'] == 'Senator'].pivot_table(
    index='choice', columns='MF', values='vote_count', fill_value=0
).rename(columns={'M': 'Guys', 'F': 'Girls'}).reset_index()

# Display results
results = {
    "Senior vs Junior Average Party Rankings": senior_junior_rankings,
    "Guys vs Girls Average Party Rankings": gender_rankings,
    "Senior vs Junior President Choices": presidential_fptp_class,
    "Senior vs Junior Senate Choices": senate_fptp_class,
    "Guys vs Girls President Choices": presidential_fptp_gender,
    "Senior vs Junior Senate Choices": senate_fptp_gender
}

for key, value in results.items():
    print(f"\n{key}:\n", value)
